java-checks/src/main/java/org/sonar/java/checks/AssertionsCompletenessCheck.java

```diff
@@ -0,0 +1,127 @@
+
+  private static final MethodInvocationMatcher MOCKITO_VERIFY = MethodInvocationMatcher.create()
+    .typeDefinition("org.mockito.Mockito").name("verify").withNoParameterConstraint();
+  private static final MethodInvocationMatcherCollection FEST_ASSERT_THAT = MethodInvocationMatcherCollection.create(
+    // Fest 1.X
+    MethodInvocationMatcher.create().typeDefinition("org.fest.assertions.Assertions").name("assertThat").addParameter(TypeCriteria.anyType()),
+    // Fest 2.X
+    MethodInvocationMatcher.create().typeDefinition("org.fest.assertions.api.Assertions").name("assertThat").addParameter(TypeCriteria.anyType())
+  );
+  private static final MethodInvocationMatcherCollection FEST_ASSERTIONS = MethodInvocationMatcherCollection.create(
+    // Fest 1.X
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.fest.assertions.GenericAssert")).name(NameCriteria.any()).withNoParameterConstraint(),
+    // Fest 2.X
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf("org.fest.assertions.api.AbstractAssert")).name(NameCriteria.any()).withNoParameterConstraint()
+  );
+
+  private static final MethodInvocationMatcherCollection FEST_EXCLUSIONS = MethodInvocationMatcherCollection.create(
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.anyType()).name("as").withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.anyType()).name("describedAs").withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.anyType()).name("overridingErrorMessage").withNoParameterConstraint()
+  );
+
+  private final Deque<ChainedMethods> chainedTo = new LinkedList<>();
+  private JavaFileScannerContext context;
+
+  private enum ChainedMethods {
+    ANY,
+    ANY_BUT_NOT_FEST_EXCLUSIONS,
+    FEST
+  }
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    // skip variable assignments
+  }
+
+  @Override
+  public void visitReturnStatement(ReturnStatementTree tree) {
+    // skip return statements
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree mit) {
+    if (maybeRaiseIssue(mit)) {
+      return;
+    }
+    if (FEST_ASSERTIONS.anyMatch(mit) && !FEST_EXCLUSIONS.anyMatch(mit)) {
+      chainedTo.push(ChainedMethods.FEST);
+    } else if (!FEST_EXCLUSIONS.anyMatch(mit)) {
+      chainedTo.push(ChainedMethods.ANY_BUT_NOT_FEST_EXCLUSIONS);
+    } else {
+      chainedTo.push(ChainedMethods.ANY);
+    }
+    super.visitMethodInvocation(mit);
+    chainedTo.pop();
+  }
+
+  private boolean maybeRaiseIssue(MethodInvocationTree mit) {
+    if (FEST_ASSERT_THAT.anyMatch(mit) && mit.arguments().size() == 1) {
+      if (chainedTo.isEmpty() || !(chainedTo.contains(ChainedMethods.FEST) || chainedTo.contains(ChainedMethods.ANY_BUT_NOT_FEST_EXCLUSIONS))) {
```

Comment:
As far as I can see using your `enum` is a bit overkill at that point. `FEST` and `ANY_BUT_NOT_FEST_EXCLUSIONS` have the same effect. That can be simplified.